File: jvm/src/zalando/CalculateStatistics.java
package zalando;

// you can also use imports, for example:
import java.util.*;


class Statistic {
    Double average;
    Integer totalNumber;

    Statistic(Double average, Integer totalNumber) {
        // TODO :average -> totalExecutionTIme more better
        this.average = average;
        this.totalNumber = totalNumber;
    }

    Double getAverage() {
        return this.average;
    }

    Integer getTotalNumber() {
        return this.totalNumber;
    }

    @Override
    public String toString() {
        return "Statistic{" +
                "average=" + average +
                ", totalNumber=" + totalNumber +
                '}';
    }
}

public class CalculateStatistics {
    /**
     *
     * HashMap
     *      Key - Method name e.g. "Method_A"
     *      Value - Statistic {
     average: 1.5, -> totalExecutionTime
     totalNumber: 2,
     }

     Stack<String>
     e.g. "{timestamp};{methodName}" - "0;Method_A"

     */
    private static void solution(String[] input) {
        // init
        final Map<String, Statistic> mem = new HashMap<>();
        final Stack<String> stk = new Stack<>();

        // iterate
        for (String line : input) {
            // manipulate string
            final String[] s = line.split(";");
            // Enter or Exit
            final String t = s[0];
            final String o = s[1];

            // determine whether Enter or Exit
            if (o.equals("Enter")) {
                // case of Enter
                final String n = s[2];
                // "{timestamp};{methodName}"
                final String sd = String.format("%s;%s", t, n);
                stk.add(sd);
            } else {
                // case of Exit
                final String[] sd = stk.pop().split(";");
                // TODO: need to parseDouble
                final Double executionTime = Double.parseDouble(t) - Double.parseDouble(sd[0]);
                final String methodName = sd[1];

                final Statistic st = mem.getOrDefault(methodName, null);
                if (st == null) {
                    mem.put(methodName, new Statistic(executionTime, 1));
                } else {


                    // calculate average
                    final Integer newNum = st.getTotalNumber() + 1;
                    final Double newAvg = ((st.getAverage() * st.getTotalNumber()) + executionTime) / newNum.doubleValue();

                    mem.put(methodName, new Statistic(newAvg, newNum));
                    //  (prevAvg * prevNum + executiontime) / (prevNum + 1)
                }
            }



        }


        // output
        System.out.println(mem);
    }



    public static void main(String [] args) {
        String[] input = {
                "0;Enter;Method_A",
                "10;Enter;Method_B",
                "15;Exit",
                "17;Exit",
                "18;Enter;Method_C",
                "22;Exit",
                "23;Enter;Method_B",
                "25;Enter;Method_A",
                "31;Exit",
                "33;Exit"
        };
        solution(input);

    }
}

File: scala/src/main/scala/zalando/CalculateStatistics.scala
package zalando

import zalando.CalculateStatistics.{ENTER, EXIT, SPLITTER}

class CalculateStatistics {
  /*
    preprocessing
    String => List[String] => List[Log]

    analyze
    List[Log] => List[Statistics]
   */
  type Log = (Int, String, Option[String])

  def preprocess(xs: List[String]): List[Log] =
    xs.map(
      _.split(SPLITTER) match {
        case Array(t, b, m) => (t.toInt, b, Option(m))
        case Array(t, b) => (t.toInt, b, Option.empty)
      }
    )

  /*
      Enter -> push
      Exit -> pop and add Statistics

      Map: method name -> statistics
   */
  def analyze(xs: List[Log]): Map[String, Statistics] = {
    @annotation.tailrec
    def loop(xs: List[Log], stk: List[Log], acc: Map[String, Statistics]): Map[String, Statistics] = {
      xs match {
        case Nil => acc
        case h :: t => h match {
          // push
          case (_, ENTER, _) => loop(t, h :: stk, acc)
          case (ext, EXIT, None) =>
            stk match {
              // pop and accumulate statistics
              case (ent, ENTER, Some(m)) :: st =>
                acc.get(m) match {
                  case Some(s) => loop(t, st, acc.updated(m, Statistics(m, s.totalTime + ext - ent, s.totalNumber + 1)))
                  case None => loop(t, st, acc.updated(m, Statistics(m, ext - ent, 1)))
                }
              case _ => throw new IllegalArgumentException()
            }
        }
        case _ => throw new IllegalArgumentException()
      }
    }

    loop(xs, List(), Map.empty)
  }


}

case class Statistics(methodName: String, totalTime: Double = 0, totalNumber: Double = 0) {
  def average: Double = totalTime / totalNumber
}

object CalculateStatistics {
  // better to be enum
  val ENTER = "Enter"
  val EXIT = "Exit"

  val SPLITTER = ";"

  /*
    input
    "0;Enter;Method_A",
    "10;Enter;Method_B",
    "15;Exit",
    "17;Exit",
    "18;Enter;Method_C",
    "22;Exit",
    "23;Enter;Method_B",
    "25;Enter;Method_A",
    "31;Exit",
    "33;Exit"
   */
  def main(args: Array[String]): Unit = {
    val cs = new CalculateStatistics()
    val p = cs.preprocess(
      List(
        "0;Enter;Method_A",
        "10;Enter;Method_B",
        "15;Exit",
        "17;Exit",
        "18;Enter;Method_C",
        "22;Exit",
        "23;Enter;Method_B",
        "25;Enter;Method_A",
        "31;Exit",
        "33;Exit"
      )
    )

    val m = cs.analyze(p)
    println(m)
    m.foreach(s => println(s"${s._1} avg: ${s._2.average}"))
  }
}
